core/src/main/java/org/apache/iceberg/BaseFileScanTask.java

```diff
@@ -86,16 +89,38 @@ public String toString() {
         .toString();
   }
 
-  /**
-   * Visible for Testing
-   */
-  static final class SplitScanTaskIterator implements Iterator<FileScanTask> {
+  @VisibleForTesting
+  static final class OffsetsBasedSplitScanTaskIterator implements Iterator<FileScanTask> {
+    private final List<Long> splitOffsets;
+    private final FileScanTask parentScanTask;
+    private int idx = 0;
+
+    OffsetsBasedSplitScanTaskIterator(List<Long> splitOffsets, FileScanTask fileScanTask) {
+      this.splitOffsets = splitOffsets;
+      this.parentScanTask = fileScanTask;
+    }
+
+    @Override
+    public boolean hasNext() {
+      return idx < splitOffsets.size();
+    }
+
+    @Override
+    public FileScanTask next() {
+      long start = splitOffsets.get(idx++);
```

Comment:
Nit: I don't think it is a good practice to use the return value of the `++` operator. It is difficult to remember whether the returned value is the original or incremented one. I think it is more clear to put this on a separate line.